src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,713 @@
+
+// Timeout for compile/link of a DLL
+const int kBuildTimeoutSecs = 120;
+
+// A quick-and-dirty way to get a hash of two strings, take 80bits,
+// and produce a 20byte string of hex digits.
+std::string hash_two_strings(const std::string& a, const std::string& b) {
+  auto hashA = std::hash<std::string>()(a);
+  auto hashB = std::hash<std::string>()(b);
+  hashA ^= (0x3ac5*hashB);
+  hashB &= 0xffff;
+  char buf[128];
+  char* out = buf;
+  for (int pos = 80; (pos -= 4) >= 0;) {
+    int64_t nibble = ((pos >= 64) ? (hashB >> (pos-64)) : (hashA >> pos)) & 0xf;
+    *out++ = ((nibble < 10) ? nibble+'0' : nibble-10+'a');
+  }
+  *out = 0;
+  return std::string(buf);
+}
+
+bool file_stat(const std::string& name, struct stat* st) {
+  // Open file for reading to avoid inconsistent cached attributes
+  int fd = ::open(name.c_str(), O_RDONLY, 0666);
+  if (fd < 0) return false;
+  int rc;
+  do {
+    errno = 0;
+    rc = ::fstat(fd, st);
+  } while ((rc < 0) && (errno == EINTR));
+  ::close(fd);
+  return (rc == 0);
+}
+
+bool file_exists_and_is_recent(const std::string& name) {
+  time_t now = ::time(nullptr);
+  struct stat st;
+  return (file_stat(name, &st) && (st.st_ctime+kBuildTimeoutSecs > now));
+}
+
+bool file_exists(const std::string& name) {
+  struct stat st;
+  return file_stat(name, &st);
+}
+
+std::string read_file_as_string(const std::string& name) {
+  FILE* f = fopen(name.c_str(), "r");
+  if (!f) return std::string("");
+  std::stringstream ss;
+  for (;;) {
+    int c = fgetc(f);
+    if (c == EOF) break;
+    ss << (char)c;
+  }
+  fclose(f);
+  return ss.str();
+}
+
+std::string getenv_with_default(const char* name, const char* dval) {
+  const char* s = ::getenv(name);
+  return std::string(s ? s : dval);
+}
+
+std::string run_shell_get_first_line(const char* cmd) {
+  FILE* f = popen(cmd, "r");
+  char buf[1024];
+  size_t len = 0;
+  if (f) {
+    for (;;) {
+      int c = fgetc(f);
+      if (c == EOF) break;
+      if ((c != '\n') && (len+1 < sizeof(buf))) buf[len++] = c;
+    }
+    pclose(f);
+  }
+  buf[len] = 0;
+  return std::string(buf);
+}
+
+std::string get_java_home() {
+  auto s = run_shell_get_first_line(
+    "java -XshowSettings:properties -version 2>&1 | fgrep -i java.home"
+  );
+  auto p = strstr(s.c_str(), "java.home = ");
+  if (p) {
+    return std::string(p+12);
+  } else {
+    return std::string(getenv_with_default("JAVA_HOME", "JAVA_HOME_undefined"));
+  }
+}
+
+std::string strip_suffix(const std::string& s, const char* suffix) {
+  size_t len = s.length();
+  size_t n = strlen(suffix);
+  if ((n > len) || (strncmp(&s[len-n], suffix, n) != 0)) return s;
+  return std::string(s, 0, len-n);
+}
+
+std::string get_cxx_name() {
+  char* p = ::getenv("CXX");
+  if (p) return std::string(p);
+  // We prefer clang because it has faster compile
+  auto s = run_shell_get_first_line("which clang");
+  if (strstr(s.c_str(), "clang")) return s;
+  s = run_shell_get_first_line("which g++");
```

Comment:
OK.  You want it to behave like a "standard" makefile in finding compiler, either taking $(CXX)
or if that's not defined then finding "c++" on $(PATH).  Then do our best to warn if we *know*
it can't work, or if flags are bad.  But then go ahead and make best effort.

I'll do it that way.

The cached Makefile isn't actually reused.  If NativeModule decides it needs to rebuild, it
always write the hm_whatever.mak.  It's occasionally useful to run it by hand for debugging.

Reply:
I'm lazy and I want to save my thinking for stuff where I can really provide value (like building a whole stage code generator, or designing a new format to speed up the decoder!)  If there are standards, I want to use them so I don't have to think unless I have a strong reason to think they won't work for me.  It also makes communicating with others easier who are expecting the same standards.  So I'd just kick out a Makefile and call make on it like I would for any project (and maybe print out some helpful info when it fails like the make command, the path to the Makefile and maybe the environment).  I didn't realize it would be such a hard sell because I thought I was proposing the easy path which was the least work.  I like the easy path!  And if it means we can get this in and start building on it, all the better.

 > The cached Makefile isn't actually reused.  ...  It's occasionally useful to run it by hand for debugging.

Ah, thanks for clarifying.  Definitely.